jep/401/README.adoc

```diff
@@ -0,0 +1,245 @@
+* Do not discuss alternative designs that were rejected - those belong in the Reasoning section.
+* Avoid in-depth discussion or justification of design choices - that belongs in the Reasoning section.
+====
+
+[]
+====
+Angelique - need to add additional info about Jenkins.READ?
+====
+
+The following permission types are now marked as `@Deprecated` and `@Restricted(NoExternalUse.class)`:
+
+* `Jenkins.RUN_SCRIPTS`
+* `Jenkins.CONFIGURE_UPDATECENTER`
+* `Jenkins.UPLOAD_PLUGINS`
+
+This allows for backwards compatiblilty with plugins while providing motivation to refactor any usages of the deprecated checks in future releases.
+
+Actions that formerly checked for these permissions will instead check for `Jenkins.ADMINISTER`.
+
+All other actions that check for `Jenkins.ADMINISTER` will be evaluated to determine if the use of the new `Jenkins.CONFIGURE` permission type is appropriate.  The general rule of thumb is that actions which do not let you change the underlying security model (either directly, or through a malicious plugin or groovy script), or change fundamental configurations such as hostnames, proxy configuration, etc are likely candidates to use the `Jenkins.CONFIGURE`.
+
+[]
+====
+do we need to talk about the matrix auth plugin here as well?  The new permissions are only applicable if using that so i think we should talk about it.
+====
+
+== Motivation
+
+[TIP]
+====
+Explain why the existing code base or process is inadequate to address the problem that the JEP solves.
+This section may also contain any historical context such as how things were done before this proposal.
+
+* Do not discuss design choices or alternative designs that were rejected - those belong in the Reasoning section.
+====
+
+The current permission model provides a false sense of security for Jenkins administrators.  The motivation was to allow some users the ability to configure limited aspects of a Jenkins instance without granting them full `Jenkins.ADMINISTER` privileges; however, it is trivial for a motivated user to grant themselves the additional privileges by means of the groovy console (`Jenkins.RUN_SCRIPTS`), or by means of a malicious plugin (`Jenkins.UPLOAD_PLUGINS`, `Jenkins.CONFIGURE_UPDATECENTER`).
+
+By consolidating all permission types that effectively allow a user to have full `root` access on a Jenkins instance, and introducing a new permission (`Jenkins.CONFIGURE`) that allows a limited amount of access to configure certain non-critical functionality, a Jenkins administrator can safely delegate configuration aspects while being confident that security concerns are being met.
+
+The `Jenkins.CONFIGURE` permission type is not intended to replace any `Item` level permissions.  It is solely for the purpose of seperating sensistive, security related Jenkins configuration settings from those that a non-root administrator might be expected to manage.
```

Comment:
Will the `configuration-as-code` plugin allow someone with only `CONFIGURE` to set those things?

Reply:
CasC sets the config as SYSTEM during start up, so there is no security check at all there. If you mean after Jenkins has started, CasC is not interacting in any way at that point (at least until the read only UI lands).

Reply:
I would hope that to use CasC you need Administrator - otherwise you could set the security realm, grant yourself Administer or whatever you wanted...
CasC itself is all-or nothing and is not fine grained enough to support configuring X vs Y.  However I expect that to happen as part of code review process for the code (if today someone has delegated administration to someone via CasC then they have delegated full permissions to that user in any case, they just need to get the code change done and where it is expected and then wait for the next restart of Jenkins)